File: 2023_8/baekjoon/2110.py
N, C = map(int, input().split())
router = []
for i in range(N):
    router.append(int(input()))
router.sort()
result = 0


def solution(start, end) -> int:
    while start <= end:
        count = 1
        mid = (start + end) // 2
        last_router_pos = router[0]

        for i in range(1, N):
            if router[i] - last_router_pos >= mid:
                count += 1
                last_router_pos = router[i]

        if count >= C:
            global result
            result = mid
            start = mid + 1
        else:
            end = mid - 1

    return result


print(solution(1, router[-1] - router[0]))

File: 2023_8/baekjoon/2110.rs
use std::io::{stdin, Read};

fn main() {
    let mut buf = String::new();
    stdin().read_to_string(&mut buf).unwrap();
    let mut buf = buf.split_ascii_whitespace();
    let mut iterator = || buf.next().unwrap().parse::<usize>().unwrap();
    let (n, c) = (iterator(), iterator());
    let mut routers: Vec<usize> = Vec::new();
    for _ in 0..n {
        routers.push(iterator());
    }
    routers.sort();
    println!(
        "{}",
        solution(1, routers.last().unwrap() - routers[0], &routers, c)
    );
}

fn solution(mut start: usize, mut end: usize, routers: &Vec<usize>, target: usize) -> usize {
    let mut result = 0;
    while start <= end {
        let mid = ((start + end) / 2) as usize;
        let mut count = 1;
        let mut last_pos = routers[0];

        for next in routers.iter().skip(1) {
            if next - last_pos >= mid {
                count += 1;
                last_pos = *next;
            }
        }

        if count >= target {
            start = mid + 1;
            result = mid;
        } else {
            end = mid - 1;
        }
    }
    result
}
